src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKeySynchronizer.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class AccessKeySynchronizer : IAccessKeySynchronizer, IDisposable
+    {
+        private readonly ConcurrentDictionary<ServiceEndpoint, object> _endpoints = new ConcurrentDictionary<ServiceEndpoint, object>(ReferenceEqualityComparer.Instance);
+
+        private readonly ILoggerFactory _factory;
+        private readonly IServerNameProvider _provider;
+        private readonly TimerAwaitable _timer = new TimerAwaitable(TimeSpan.Zero, TimeSpan.FromMinutes(5));
+
+        public AccessKeySynchronizer(
+            IServerNameProvider nameProvider,
+            ILoggerFactory loggerFactory
+            )
+        {
+            _ = UpdateAsync();
+
+            _provider = nameProvider;
+            _factory = loggerFactory;
+        }
+
+        /// <summary>
+        /// For test only.
+        /// </summary>
+        private AccessKeySynchronizer()
+        {
+            _provider = new DefaultServerNameProvider();
+            _factory = NullLoggerFactory.Instance;
+        }
+
+        internal static AccessKeySynchronizer ForTest() => new AccessKeySynchronizer();
```

Comment:
Do not add this ForTest instance in production code.

```cs
// for tests
internal AccessKeySynchronizer(IServerNameProvider nameProvider, ILoggerFactory loggerFactory)
{
   _provider = nameProvider;
   _factory = loggerFactory;
}
```
And use in test
```
new AccessKeySynchronizer(new DefaultServerNameProvider(), NullLoggerFactory.Instance);
```

Reply:
I think giving a static method with `ForTest` name and leave the constructor private could be better than just comment it as `for tests`?

Reply:
I think having these ForTest instance in production code is quite incompliant. Please follow other class about how to inherit or use in tests.